File: server/routes/tickets.js
const express = require("express");
const prisma = require("../db");
const { sendTicketCreatedWebhook } = require("../services/webhook");
const {
  VALID_CHANNELS,
  VALID_STATUSES,
  VALID_COMPLIANCE_STATUSES,
  VALID_TYPES,
  VALID_PRIORITIES,
  toArray,
  serializeTicket,
} = require("../ticketSerializer");

const router = express.Router();

function badRequest(res, message) {
  return res.status(400).json({ error: { title: "InvalidRecord", message } });
}

// POST /api/v2/tickets.json
router.post("/tickets.json", async (req, res, next) => {
  try {
    const body = req.body.ticket || req.body;

    if (!body || !body.subject || !body.description) {
      return badRequest(res, "subject and description are required");
    }
    if (!body.requester_email || !body.requester_name) {
      return badRequest(res, "requester_name and requester_email are required");
    }
    const channel = body.channel || "web_form";
    if (!VALID_CHANNELS.includes(channel)) {
      return badRequest(res, `channel must be one of ${VALID_CHANNELS.join(", ")}`);
    }
    if (body.type !== undefined && body.type !== null && !VALID_TYPES.includes(body.type)) {
      return badRequest(res, `type must be one of ${VALID_TYPES.join(", ")}`);
    }
    if (body.priority !== undefined && !VALID_PRIORITIES.includes(body.priority)) {
      return badRequest(res, `priority must be one of ${VALID_PRIORITIES.join(", ")}`);
    }

    const ticket = await prisma.ticket.create({
      data: {
        subject: body.subject,
        description: body.description,
        channel,
        requesterName: body.requester_name,
        requesterEmail: body.requester_email,
        status: VALID_STATUSES.includes(body.status) ? body.status : "new",
        type: body.type ?? null,
        priority: VALID_PRIORITIES.includes(body.priority) ? body.priority : "normal",
        groupName: body.group_name ?? null,
        assignee: body.assignee ?? null,
        tags: toArray(body.tags).join(","),
        classificationResult: body.classification_result ?? null,
        confidenceScore: body.confidence_score ?? null,
        complianceStatus: VALID_COMPLIANCE_STATUSES.includes(body.compliance_status)
          ? body.compliance_status
          : "none",
        flagType: JSON.stringify(toArray(body.flag_type)),
      },
    });

    // Fire-and-forget: do not block the response on webhook delivery.
    sendTicketCreatedWebhook(ticket).catch(() => {});

    res.status(201).json({ ticket: serializeTicket(ticket) });
  } catch (err) {
    next(err);
  }
});

// GET /api/v2/tickets.json
router.get("/tickets.json", async (req, res, next) => {
  try {
    const { status, group_name, flag_type } = req.query;
    const where = {};

    if (status) where.status = status;
    if (group_name) where.groupName = group_name;
    if (flag_type) where.flagType = { contains: `"${flag_type}"` };

    const tickets = await prisma.ticket.findMany({
      where,
      orderBy: { createdAt: "desc" },
    });

    res.json({ count: tickets.length, tickets: tickets.map((t) => serializeTicket(t)) });
  } catch (err) {
    next(err);
  }
});

// GET /api/v2/tickets/:id.json
router.get("/tickets/:id.json", async (req, res, next) => {
  try {
    const id = Number(req.params.id);
    const ticket = await prisma.ticket.findUnique({
      where: { id },
      include: { notes: { orderBy: { createdAt: "asc" } } },
    });

    if (!ticket) {
      return res.status(404).json({ error: { title: "RecordNotFound", message: "Ticket not found" } });
    }

    res.json({ ticket: serializeTicket(ticket, { includeNotes: true }) });
  } catch (err) {
    next(err);
  }
});

// PUT /api/v2/tickets/:id.json
router.put("/tickets/:id.json", async (req, res, next) => {
  try {
    const id = Number(req.params.id);
    const existing = await prisma.ticket.findUnique({ where: { id } });
    if (!existing) {
      return res.status(404).json({ error: { title: "RecordNotFound", message: "Ticket not found" } });
    }

    const body = req.body.ticket || req.body;
    const data = {};

    if (body.status !== undefined) {
      if (!VALID_STATUSES.includes(body.status)) {
        return badRequest(res, `status must be one of ${VALID_STATUSES.join(", ")}`);
      }
      data.status = body.status;
    }
    if (body.type !== undefined) {
      if (body.type !== null && !VALID_TYPES.includes(body.type)) {
        return badRequest(res, `type must be one of ${VALID_TYPES.join(", ")}`);
      }
      data.type = body.type;
    }
    if (body.priority !== undefined) {
      if (!VALID_PRIORITIES.includes(body.priority)) {
        return badRequest(res, `priority must be one of ${VALID_PRIORITIES.join(", ")}`);
      }
      data.priority = body.priority;
    }
    if (body.group_name !== undefined) data.groupName = body.group_name;
    if (body.assignee !== undefined) data.assignee = body.assignee;
    if (body.tags !== undefined) data.tags = toArray(body.tags).join(",");
    if (body.classification_result !== undefined) data.classificationResult = body.classification_result;
    if (body.confidence_score !== undefined) data.confidenceScore = body.confidence_score;
    if (body.compliance_status !== undefined) {
      if (!VALID_COMPLIANCE_STATUSES.includes(body.compliance_status)) {
        return badRequest(res, `compliance_status must be one of ${VALID_COMPLIANCE_STATUSES.join(", ")}`);
      }
      data.complianceStatus = body.compliance_status;
    }
    if (body.flag_type !== undefined) data.flagType = JSON.stringify(toArray(body.flag_type));
    if (body.subject !== undefined) data.subject = body.subject;
    if (body.description !== undefined) data.description = body.description;

    const ticket = await prisma.ticket.update({ where: { id }, data });

    res.json({ ticket: serializeTicket(ticket) });
  } catch (err) {
    next(err);
  }
});

// POST /api/v2/tickets/:id/notes.json
router.post("/tickets/:id/notes.json", async (req, res, next) => {
  try {
    const id = Number(req.params.id);
    const ticket = await prisma.ticket.findUnique({ where: { id } });
    if (!ticket) {
      return res.status(404).json({ error: { title: "RecordNotFound", message: "Ticket not found" } });
    }

    const body = req.body.note || req.body;
    if (!body || !body.text) {
      return badRequest(res, "text is required");
    }

    const note = await prisma.internalNote.create({
      data: {
        ticketId: id,
        text: body.text,
        createdBy: body.created_by || "system",
      },
    });

    res.status(201).json({
      note: {
        id: note.id,
        ticket_id: note.ticketId,
        text: note.text,
        created_by: note.createdBy,
        created_at: note.createdAt,
      },
    });
  } catch (err) {
    next(err);
  }
});

module.exports = router;

File: fastapi-app/app/routers/tickets.py
import json
from datetime import datetime, timezone

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException
from sqlmodel import Session, select

from ..database import get_session
from ..deps import require_auth
from ..models import InternalNote, Ticket
from ..schemas import (
    NoteCreateWrapper,
    TicketCreateWrapper,
    TicketUpdateWrapper,
    normalize_list,
    to_csv,
)
from ..serializers import serialize_ticket
from ..services.webhook import send_ticket_created_webhook

router = APIRouter(prefix="/api/v2", dependencies=[Depends(require_auth)])


def _not_found():
    return HTTPException(status_code=404, detail={"title": "RecordNotFound", "message": "Ticket not found"})


@router.post("/tickets.json", status_code=201)
def create_ticket(
    body: TicketCreateWrapper,
    background_tasks: BackgroundTasks,
    session: Session = Depends(get_session),
):
    data = body.ticket
    ticket = Ticket(
        subject=data.subject,
        description=data.description,
        channel=data.channel,
        requester_name=data.requester_name,
        requester_email=data.requester_email,
        status=data.status,
        type=data.type,
        priority=data.priority,
        group_name=data.group_name,
        assignee=data.assignee,
        tags=to_csv(data.tags),
        classification_result=data.classification_result,
        confidence_score=data.confidence_score,
        compliance_status=data.compliance_status,
        flag_type=json.dumps(normalize_list(data.flag_type)),
    )
    session.add(ticket)
    session.commit()
    session.refresh(ticket)

    serialized = serialize_ticket(ticket)
    # Fire-and-forget: scheduled to run after the response is sent, so it
    # never delays this request.
    background_tasks.add_task(send_ticket_created_webhook, serialized)

    return {"ticket": serialized}


@router.get("/tickets.json")
def list_tickets(
    status: str | None = None,
    group_name: str | None = None,
    flag_type: str | None = None,
    session: Session = Depends(get_session),
):
    query = select(Ticket)
    if status:
        query = query.where(Ticket.status == status)
    if group_name:
        query = query.where(Ticket.group_name == group_name)
    if flag_type:
        query = query.where(Ticket.flag_type.contains(f'"{flag_type}"'))
    query = query.order_by(Ticket.created_at.desc())

    tickets = session.exec(query).all()
    return {"count": len(tickets), "tickets": [serialize_ticket(t) for t in tickets]}


@router.get("/tickets/{ticket_id}.json")
def get_ticket(ticket_id: int, session: Session = Depends(get_session)):
    ticket = session.get(Ticket, ticket_id)
    if not ticket:
        raise _not_found()
    return {"ticket": serialize_ticket(ticket, include_notes=True)}


@router.put("/tickets/{ticket_id}.json")
def update_ticket(
    ticket_id: int,
    body: TicketUpdateWrapper,
    session: Session = Depends(get_session),
):
    ticket = session.get(Ticket, ticket_id)
    if not ticket:
        raise _not_found()

    updates = body.ticket.model_dump(exclude_unset=True)

    if "tags" in updates:
        updates["tags"] = to_csv(updates["tags"])
    if "flag_type" in updates:
        updates["flag_type"] = json.dumps(normalize_list(updates["flag_type"]))

    for field, value in updates.items():
        setattr(ticket, field, value)
    ticket.updated_at = datetime.now(timezone.utc)

    session.add(ticket)
    session.commit()
    session.refresh(ticket)

    return {"ticket": serialize_ticket(ticket)}


@router.post("/tickets/{ticket_id}/notes.json", status_code=201)
def add_note(
    ticket_id: int,
    body: NoteCreateWrapper,
    session: Session = Depends(get_session),
):
    ticket = session.get(Ticket, ticket_id)
    if not ticket:
        raise _not_found()

    note = InternalNote(ticket_id=ticket_id, text=body.note.text, created_by=body.note.created_by)
    session.add(note)
    session.commit()
    session.refresh(note)

    return {
        "note": {
            "id": note.id,
            "ticket_id": note.ticket_id,
            "text": note.text,
            "created_by": note.created_by,
            "created_at": note.created_at,
        }
    }
